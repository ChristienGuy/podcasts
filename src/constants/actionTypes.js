// Login
export const AUTHENTICATION_LOGIN_ATTEMPT: string = "AUTHENTICATION_LOGIN_ATTEMPT"
export const AUTHENTICATION_LOGIN_FAILURE: string = "AUTHENTICATION_LOGIN_FAILURE"
export const AUTHENTICATION_LOGIN_SUCCESS: string = "AUTHENTICATION_LOGIN_SUCCESS"

// Logout
export const AUTHENTICATION_LOGOUT_FAILURE: string = "AUTHENTICATION_LOGOUT_FAILURE"
export const AUTHENTICATION_LOGOUT_SUCCESS: string = "AUTHENTICATION_LOGOUT_SUCCESS"

// Session
export const AUTHENTICATION_SESSION_CHECK_ATTEMPT: string = "AUTHENTICATION_SESSION_CHECK_ATTEMPT"
export const AUTHENTICATION_SESSION_CHECK_FAILURE: string = "AUTHENTICATION_SESSION_CHECK_FAILURE"
export const AUTHENTICATION_SESSION_CHECK_SUCCESS: string = "AUTHENTICATION_SESSION_CHECK_SUCCESS"


// Podcast
export const PODCAST_SET_PODCASTS: string = "PODCAST_SET_PODCASTS"
export const PODCAST_ADD_PODCAST: string = "PODCAST_ADD_PODCASTS"
export const PODCAST_UPDATE_PODCASTS_ATTEMPT: string = "PODCAST_UPDATE_PODCASTS_ATTEMPT"