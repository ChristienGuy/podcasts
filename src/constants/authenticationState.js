/* 
Tracking states via "enum" rather than boolean flags

Boolean flags lead to potentailly unexpected combinations of states
e.g

isLoggedIn = true
isLoggingIn = false
isCheckingSession = false
isLoadingPodcasts = false
podcastsLoaded = true

This creates 5! number of potential states (1 * 2 * 3 * 4 * 5 = 120)

See: https://christienguy.co.uk/post/react-state-enum/ for full info
*/

// 12 possible combinations of state (3 * 4) is more manageable but still not great
export const AuthenticationState = Object.freeze({
  LOGGING_IN: "LOGGING_IN",
  LOGGED_IN: "LOGGED_IN",
  CHECKING_SESSION: "CHECKING_SESSION",
  NOT_AUTHENTICATED: "NOT_AUTHENTICATED",
})
