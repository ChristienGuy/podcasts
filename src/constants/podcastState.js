export const PodcastState = Object.freeze({
  UPDATING: "UPDATING",
  LOADING: "LOADING",
  LOADED: "LOADED",
  NOT_LOADED: "NOT_LOADED"
})