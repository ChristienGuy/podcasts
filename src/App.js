import React, { Component, Fragment } from "react";
import { bindActionCreators } from "redux";
import { connect } from "react-redux";
import axios from "axios";

import {
  sessionCheckFailure,
  sessionCheckSuccess,
  sessionCheckAttempt
} from "./actions/authentication";

import { AuthenticationState } from "./constants/authenticationState";

import { setPodcasts, updatePodcastsAttempt } from "./actions/podcasts";

import { BASE_URL } from "./constants";

import { BrowserRouter as Router, Route, Switch } from "react-router-dom";

import Player from "./components/shared/Player";

import HeaderContainer from "./components/shared/HeaderContainer";
import LoginPageContainer from "./components/account/LoginPageContainer";
import RegisterPageContainer from "./components/account/RegisterPageContainer";
import PodcastsPageContainer from "./components/podcasts/PodcastsPageContainer";
import AddPodcastPageContainer from "./components/podcasts/AddPodcastPageContainer";
import EpisodesPageContainer from "./components/podcasts/EpisodesPageContainer";

export class App extends Component {
  componentWillMount() {
    this.checkSession();
  }

  updatePodcasts = e => {
    const { setPodcasts, updatePodcastsAttempt } = this.props;

    updatePodcastsAttempt();
    axios({
      method: "GET",
      url: `${BASE_URL}/api/podcast/update`,
      withCredentials: true
    })
      .then(response => {
        if (response.status === 200) {
          return response.data;
        }
        return null;
      })
      .then(podcasts => {
        setPodcasts(podcasts);
      });
  };

  checkSession = async () => {
    const {
      sessionCheckFailure,
      sessionCheckSuccess,
      sessionCheckAttempt,
      setPodcasts
    } = this.props;

    sessionCheckAttempt()

    await axios({
      url: `${BASE_URL}/api/authentication/checksession`,
      method: "GET",
      responseType: "json",
      withCredentials: true
    })
      .then(response => {
        if (response.status === 200) {
          return response.data;
        }
        return null;
      })
      .then(json => {
        if (json.username) {
          sessionCheckSuccess(json);
          setPodcasts(json.podcasts);
          // session is checked succesfully, go check for updates to podcasts
          this.updatePodcasts()
        } else {
          sessionCheckFailure();
        }
      })
      .catch(err => {
        sessionCheckFailure(err);
      });
  };

  render() {
    return (
      <Router>
        <Fragment>
          <HeaderContainer />
          {this.props.authState === AuthenticationState.CHECKING_SESSION &&
            <p>Loading...</p>
          }
          {this.props.authState === AuthenticationState.LOGGED_IN && 
            <Fragment>
              <Switch>
                <Route path="/login" component={LoginPageContainer} />
                <Route path="/register" component={RegisterPageContainer} />
                <Route
                  path="/podcast/add"
                  component={AddPodcastPageContainer}
                />
                <Route
                  path="/podcast/:name"
                  component={EpisodesPageContainer}
                />
                <Route path="/" component={PodcastsPageContainer} />
              </Switch>
              <Player />
            </Fragment>
          }
          {this.props.authState === AuthenticationState.NOT_AUTHENTICATED && 
            <Fragment>
              <Switch>
                <Route path="/" component={LoginPageContainer} />
                <Route path="/login" component={LoginPageContainer} />
              </Switch>
            </Fragment>
          }
        </Fragment>
      </Router>
    );
  }
}

const mapDispatchToProps = dispatch => {
  return bindActionCreators(
    {
      sessionCheckAttempt,
      sessionCheckFailure,
      sessionCheckSuccess,
      setPodcasts,
      updatePodcastsAttempt
    },
    dispatch
  );
};

const mapStateToProps = state => {
  return {
    authState: state.authentication.authenticationState
  };
};

export default connect(mapStateToProps, mapDispatchToProps)(App);
