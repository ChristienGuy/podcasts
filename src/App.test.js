import React from 'react';
import { App } from './App';
import { shallow } from "enzyme";

import { AuthenticationState } from "./constants/authenticationState";

const sharedProps = {
  sessionCheckAttempt: jest.fn(),
  sessionCheckFailure: jest.fn()
}

function setup() {
  const props = {
    sessionCheckAttempt: jest.fn(),
    sessionCheckFailure: jest.fn(),
    authState: ''
  }

  const enzymeWrapper = shallow(<App {...props} />);

  return {
    props,
    enzymeWrapper
  }
}

it('renders without crashing', () => {
  const { enzymeWrapper } = setup()
  expect(enzymeWrapper).toMatchSnapshot()
});

it("renders loading when authstate is 'CHECKING_SESSION'", () => {
  const props = {
    ...sharedProps,
    authState: AuthenticationState.CHECKING_SESSION
  }

  const wrapper = shallow(<App {...props} />)
  expect(wrapper).toMatchSnapshot()
})

it("renders correct routes when authState is 'LOGGED_IN'", () => {
  const props = {
    ...sharedProps,
    authState: AuthenticationState.LOGGED_IN
  }
  const wrapper = shallow(<App {...props} />)
  expect(wrapper).toMatchSnapshot()
})
