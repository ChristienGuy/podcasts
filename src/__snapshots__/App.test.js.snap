// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correct routes when authState is 'LOGGED_IN' 1`] = `
<BrowserRouter>
  <React.Fragment>
    <Connect(HeaderContainer) />
    <React.Fragment>
      <Switch>
        <Route
          component={[Function]}
          path="/login"
        />
        <Route
          component={[Function]}
          path="/register"
        />
        <Route
          component={[Function]}
          path="/podcast/add"
        />
        <Route
          component={[Function]}
          path="/podcast/:name"
        />
        <Route
          component={[Function]}
          path="/"
        />
      </Switch>
      <Connect(Component) />
    </React.Fragment>
  </React.Fragment>
</BrowserRouter>
`;

exports[`renders loading when authstate is 'CHECKING_SESSION' 1`] = `
<BrowserRouter>
  <React.Fragment>
    <Connect(HeaderContainer) />
    <p>
      Loading...
    </p>
  </React.Fragment>
</BrowserRouter>
`;

exports[`renders without crashing 1`] = `
<BrowserRouter>
  <React.Fragment>
    <Connect(HeaderContainer) />
  </React.Fragment>
</BrowserRouter>
`;
